### DISCUSSION TITLE
cache 的实现可能不太对，提一个建议

### DISCUSSION THREAD

**Discussion by wen-long** [IrineSistiana/mosdns#535] [General]
建议 cache 这块可以参考，或者直接用 coredns 的实现
dns cache 应该没有人直接用客户端发来的二进制做 key 的

实际影响的话，系统 api，dig，doggo 等工具对同实质内容的 DNS 请求可能会构造出不同的二进制，甚至是每次请求都不同，可能导致一直 miss

我使用上是 coredns 做 moddns 的前端来负责缓存的

**Comment by w22gb8** [comment]
修改下用q.Question[0].String()当key就行了，不包含动态cookie。

**Comment by urlesistiana** [comment]
> 实际影响的话，系统 api，dig，doggo 等工具对同实质内容的 DNS 请求可能会构造出不同的二进制，甚至是每次请求都不同，可能导致一直 miss

理论上是这样。实际上没多大影响。cache 默认不缓存 edns0。系统 api 也没有 edns0。edns0使用率极小。二进制实际上都一样。

coredns 拿qname qtype 做key，彻底不管 edns0。mosdns直接二进制做key，edns0不混就行。都是作为转发器”cache能用就行“的实现。dns cache 太复杂，有需求的不如前置 unbound。

**Reply by wen-long** [reply]
> > 实际影响的话，系统 api，dig，doggo 等工具对同实质内容的 DNS 请求可能会构造出不同的二进制，甚至是每次请求都不同，可能导致一直 miss
> 
> 理论上是这样。实际上没多大影响。cache 默认不缓存 edns0。系统 api 也没有 edns0。edns0使用率极小。二进制实际上都一样。
> 
> coredns 拿qname qtype 做key，彻底不管 edns0。mosdns直接二进制做key，edns0不混就行。都是作为转发器”cache能用就行“的实现。dns cache 太复杂，有需求的不如前置 unbound。

"实际影响的话，"
